libs/mbedtls/mbedtls_stubs.c

```diff
@@ -0,0 +1,307 @@
+#include <ctype.h>
+#include <string.h>
+#include <stdio.h>
+
+#ifdef _WIN32
+#include <windows.h>
+#include <wincrypt.h>
+#endif
+
+#include <caml/mlvalues.h>
+#include <caml/alloc.h>
+#include <caml/memory.h>
+#include <caml/fail.h>
+#include <caml/callback.h>
+
+#include "mbedtls/debug.h"
+#include "mbedtls/error.h"
+#include "mbedtls/config.h"
+#include "mbedtls/ssl.h"
+#include "mbedtls/entropy.h"
+#include "mbedtls/ctr_drbg.h"
+#include "mbedtls/net_sockets.h"
+#include "mbedtls/certs.h"
+
+#define SSL(v) (mbedtls_ssl_context*)Field(v, 0)
+#define CFG(v) (mbedtls_ssl_config*)Field(v, 0)
+#define CTR(v) (mbedtls_ctr_drbg_context*)Field(v, 0)
+#define CERT(v) (mbedtls_x509_crt*)Field(v, 0)
+#define PVOID(v) (void*)Field(v, 0)
+
+void debug(void* ctx, int debug_level, const char* file_name, int line, const char* message) {
+	printf("%s:%i: %s", file_name, line, message);
+}
+
+CAMLprim value ml_mbedtls_strerror(value code) {
+	CAMLparam1(code);
+	CAMLlocal1(r);
+	char buf[128];
+	mbedtls_strerror(Int_val(code), buf, sizeof(buf));
+	r = caml_copy_string(buf);
+	CAMLreturn(r);
+}
+
+// CtrDrbg
+
+CAMLprim value ml_mbedtls_ctr_drbg_init() {
+	CAMLparam0();
+	CAMLlocal1(obj);
+	mbedtls_ctr_drbg_context* ctr_drbg = malloc(sizeof(mbedtls_ctr_drbg_context));
+	obj = caml_alloc(1, Abstract_tag);
+	Store_field(obj, 0, (value)ctr_drbg);
+	mbedtls_ctr_drbg_init(ctr_drbg);
+	CAMLreturn(obj);
+}
+
+CAMLprim value ml_mbedtls_ctr_drbg_random(value p_rng, value output, value output_len) {
+	CAMLparam3(p_rng, output, output_len);
+	CAMLlocal1(r);
+	r = mbedtls_ctr_drbg_random(CTR(p_rng), String_val(output), Int_val(output_len));
+	CAMLreturn(Val_int(r));
```

Comment:
The `Val_int` should be around the `mbedtls_...` call – `r` should be a `value`, as declared in `CAMLlocal1`. It works anyway since they are ints too but still.

Alternatively this one (and all the ones below with the same comment) could be changed directly to `CAMLreturn(Val_int(mbedtls_...));`, saving two lines and a local declaration.